### PR TITLE
document for utils

### PR DESCRIPTION
**Pull request by MrLYG**
document for utils folder

### CODE REVIEW

File: ablator/utils/base.py [hunk 1]
```diff
@@ -61,6 +122,19 @@ def set_seed(seed: int):
 
 
 def get_lr(optimizer):
+    """
+    Get the learning rate.
```
Comment (hjzccc) [inline ablator/utils/base.py]: get the learning rate from a optimizer

File: ablator/utils/base.py [hunk 2]
```diff
@@ -70,16 +144,53 @@ def get_lr(optimizer):
...
+
+    Parameters
+    ----------
+    checkpoint_dir : Path
+        The directory containing checkpoint files.
+
+    Returns
+    -------
+    list[Path]
+        A list of the latest checkpoint files sorted by their creation time.
```
Comment (hjzccc) [inline ablator/utils/base.py]: sorted by filename, no time

File: ablator/utils/base.py [hunk 3]
```diff
@@ -70,16 +144,53 @@ def get_lr(optimizer):
...
 
 def parse_device(device: ty.Union[str, list[str]]):
+    """
+    Parse a device string or a list of device strings.
+
+    Parameters
+    ----------
+    device : ty.Union[str, list[str]]
+        The target device for the tensors.
+
```
Comment (hjzccc) [inline ablator/utils/base.py:188]: add examples for this

File: ablator/utils/file.py [hunk 4]
```diff
@@ -42,18 +90,62 @@ def default_val_parser(val):
...
+
+    Parameters
+    ----------
+    _dict : dict
+        The dictionary to update.
+    keys : list[str]
+        List of keys representing the nested path.
+    value : ty.Any
+        The value need to set at the specified path.
+
```
Comment (hjzccc) [inline ablator/utils/file.py:143]: add example here

**Review by hjzccc — COMMENTED** [review]
looks good

File: ablator/utils/file.py [hunk 5]
```diff
@@ -18,10 +33,30 @@ def make_sub_dirs(parent: str | Path, *dir_names) -> list[Path]:
 
 
 def save_checkpoint(state, filename="checkpoint.pt"):
+    """
+    Save a checkpoint of the given state.
+
+    Parameters
+    ----------
+    state : dict
+        State dictionary to save.
```
Comment (hieuchi911) [inline ablator/utils/file.py]: Maybe clarify that this is model state

File: ablator/utils/base.py [hunk 6]
```diff
@@ -25,6 +25,19 @@ def __getitem__(self, *args, **kwargs):
 
 
 def iter_to_numpy(iterable):
+    """
+    Convert torch.Tensor elements to NumPy arrays.
```
Comment (hieuchi911) [inline ablator/utils/base.py]: Not necessarily a tensor, v can be any Iterable

File: ablator/utils/base.py [hunk 7]
```diff
@@ -34,12 +47,47 @@ def iter_to_numpy(iterable):
...
     data_dict, device
 ) -> ty.Union[Sequence[torch.Tensor], dict[str, torch.Tensor]]:
+    """
+    Moving torch.Tensor elements to the specified device.
+
+    Parameters
+    ----------
+    data_dict : dict or list
+        The input dictionary or list containing torch.Tensor elements.
+    device : torch.device
```
Comment (hieuchi911) [inline ablator/utils/base.py]: According to its use in ablator (in `wrapper.to_device()`), `device` is just a string, although it can also be of type torch.device. Maybe add `str` as another possible type for this parameter

File: ablator/utils/base.py [hunk 8]
```diff
@@ -61,6 +122,19 @@ def set_seed(seed: int):
 
 
 def get_lr(optimizer):
+    """
+    Get the learning rate from a optimizer.
```
Comment (hieuchi911) [inline ablator/utils/base.py]:
```suggestion
    Get the learning rate from an optimizer.
```

File: ablator/utils/base.py [hunk 9]
```diff
@@ -70,16 +144,66 @@ def get_lr(optimizer):
...
+    -----
+    Return if the debugger is currently active
+    """
     gettrace = getattr(sys, "gettrace", lambda: None)
     return gettrace() is not None
 
 
 def get_latest_chkpts(checkpoint_dir: Path) -> list[Path]:
+    """
+    Get a list of the latest checkpoint files in a directory.
```
Comment (hieuchi911) [inline ablator/utils/base.py]: "Latest" is vague, I think it returns a list all checkpoint paths, but sorted from latest to earliest

File: ablator/utils/base.py [hunk 10]
```diff
@@ -70,16 +144,66 @@ def get_lr(optimizer):
...
+    -------
+    list[Path]
+        A list of the latest checkpoint files sorted by filename.
+    """
     return sorted(list(checkpoint_dir.glob("*.pt")))[::-1]
 
 
 def parse_device(device: ty.Union[str, list[str]]):
+    """
+    Parse a device string or a list of device strings.
```
Comment (hieuchi911) [inline ablator/utils/base.py]: - We should mention when `ValueError` is raised (when `device` a string but is not either `cuda` or `cpu` cor starts with `cuda:`)
- There's an `if` condition for the case where `device` is an `int`, but it's not included in the param type definition. Maybe raise an Issue about this?
- If `device` is not of any type defined (str, list[str], int) then this class will return `cuda` if available, otherwise `cpu`

File: ablator/utils/base.py [hunk 11]
```diff
@@ -95,6 +219,14 @@ def parse_device(device: ty.Union[str, list[str]]):
 
 
 def init_weights(module: nn.Module):
+    """
+    Initialize the weights of a module.
```
Comment (hieuchi911) [inline ablator/utils/base.py:228]: - If module is a Linear layer, initialize weight values to be coming from a normal distribution N(mu=0,std=1.0). Biases, if available, are initialized to zeros
- If module is an Embedding layer, initialize embeddings such that values in each embedding comes from N(mu=0,std=1.0). If padding is enabled, set padding embedding to a zero vector
- If module is a Normalization layer, set all biases to zeros and all weights to 1

File: ablator/utils/base.py (hunk 6 above)
Comment (MrLYG) [reply ablator/utils/base.py]: I think the logic of code is that only torch.Tensor elements will be converted to NumPy arrays.
How about "Convert elements of the input iterable to NumPy arrays if they are torch.Tensor objects."

File: ablator/utils/base.py (hunk 10 above)
Comment (MrLYG) [reply ablator/utils/base.py]: Already submit a issue.

**Review by hjzccc — APPROVED** [review]

**Review by hieuchi911 — APPROVED** [review]

File: ablator/utils/base.py [hunk 12]
```diff
@@ -25,6 +25,19 @@ def __getitem__(self, *args, **kwargs):
...
+    Convert elements of the input iterable to NumPy arrays if they are torch.Tensor objects.
+
+    Parameters
+    ----------
+    iterable : Iterable
+        The input iterable.
+
+    Returns
+    -------
+    any
```
Comment (fostiropoulos) [inline ablator/utils/base.py:38]: ty.Any*

**Review by fostiropoulos — APPROVED** [review]
